### PR TITLE
[Jest] Unit Tests for utility functions

### PR DESCRIPTION
**Pull request by Chadha93**
**Notes for Reviewers**

1. Initialises jest tests for Meshery UI.
2. Write a jest-based unit test for two utility functions.

image

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Review by Abhishek-kumar09 — APPROVED** [review]
Thanks @Chadha93 Looks Good to Me

File: ui/_tests_/utils/camel_to_sentence.test.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { CamelCaseToSentanceCase } from "../../utils/camelCaseToSentanceCase";
```
Comment (Abhishek-kumar09) [inline ui/_tests_/utils/camel_to_sentence.test.js:1]: we have a convention of naming the file like `testNameinCamelCase_spec.js`

**Review by Abhishek-kumar09 — CHANGES_REQUESTED** [review]
Frontend test is failing

**Comment by goldfishdolphin** [comment]
Looks good to me. 
I have some suggestions: 
- test some additional inputs including empty string
- test if the function is not mutating the input data

**Comment by Chadha93** [comment]
> Hey lets wrap or move forward with it. Requires some work to test different edge cases as well.

Right, wrapping with the code review.

File: ui/_tests_/utils/camel_to_sentence.test.js (hunk 1 above)
Comment (Chadha93) [reply ui/_tests_/utils/camel_to_sentence.test.js]: Gotcha.

**Comment by Chadha93** [comment]
:skull: Chanign the file name failed jest CI.

**Comment by leecalcote** [comment]
@8rxn @MarioArriaga92 thoughts to offer here, by chance?

**Comment by MarioArriaga92** [comment]
@leecalcote @Chadha93 looks like we need to improve either the jest workflow or add jest config file (`No tests found, exiting with code 1`):
CleanShot 2023-05-26 at 16 29 39@2x

**Comment by 8rxn** [comment]
Found a few typos here and there like `pascla` , are they intentional @Chadha93 ?

Also, run jest seems to not being able to locate the test files and hence the test fails. 
I thought the double "\_" as in "__" as opposed to single "_" in the folder-name could be a reason but the second expression must have covered for it I believe so this mustn't be the issue.

**Comment by Chadha93** [comment]
Hmm, I lost track of this PR.
Right @MarioArriaga92 The workflow needs to be refined, it was working with .tests in the file name. @8rxn Good catch, it's a typo, but should not affect it as it's the name of the test. @8rxn Feel free to make suggestions for the PR code.

**Review by Chadha93 — COMMENTED** [review]
@MarioArriaga92 I changed the file name to .test.js, that's how jest is reading, I didn't get how to update the config so it would read without .test. Any suggestions? The tests are passing now.

**Review by MarioArriaga92 — DISMISSED** [review]
LGTM

File: ui/_tests_/utils/testNameinCamelCase_spec.test.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { CamelCaseToSentanceCase } from "../../utils/camelCaseToSentanceCase";
+
+describe('CamelCase String Utility', () => {
+  test('make camelCase to sentansecase', () => {
+    const val = CamelCaseToSentanceCase("mesheryComponents");
```
Comment (samyakshah3008) [inline ui/_tests_/utils/testNameinCamelCase_spec.test.js:5]: Nitpik: Is it `sentanse` or `sentance` or `sentence`?

File: ui/_tests_/utils/testNameinCamelCase_spec.test.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+    expect(val).toBe("meshery  Components")
+  })
```
Comment (samyakshah3008) [inline ui/_tests_/utils/testNameinCamelCase_spec.test.js:7]: Are you sure we will have double space between meshery and Components? 
I see tests are passing so curious to know why we are having double space, is it intentional or nitpik? 🤔

File: ui/_tests_/utils/testPascletoKebabCase_spec.test.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import PascalCaseToKebab from "../../utils/PascalCaseToKebab";
+
+describe('PascalKebab String Utility', () => {
+  test('make pasclacase to kebabcase', () => {
```
Comment (samyakshah3008) [inline ui/_tests_/utils/testPascletoKebabCase_spec.test.js:4]: Nitpik: I guess in it will be `pascalcase` and not `pasclacase` 😄 🤔  
Also, if it is so, can you please correct the file name as well? It is currently `testPascletoKebabCase`

**Review by samyakshah3008 — APPROVED** [review]
Great work 💪 @Chadha93  
I have just added few single line comments, please check 😄  
Rest looks all perfect to me! 🚀

**Comment by leecalcote** [comment]
Yes, no, maybe so, @Chadha93?

**Comment by leecalcote** [comment]
Merge conflict

**Comment by Chadha93** [comment]
Oh apologies for missing this out again, will surely get on this tonight.

**Comment by l5io** [comment]
🚀 Preview for commit b36231353e8ed28a260ffe911693d7ecffb04943 at: https://649200497af7c5250b6f9043--meshery-docs-preview.netlify.app

**Comment by l5io** [comment]
🚀 Preview for commit db1e277e5546ab1cb43aad06fdf42fe3d3b53892 at: https://649202d96c2d0020fd55df6f--meshery-docs-preview.netlify.app

**Comment by Chadha93** [comment]
Uh oh, I made a mess while resolving the merge conflict, will resolve it or will make a new PR.

**Comment by Chadha93** [comment]
Fixed. Can be merged after tests pass. :heavy_check_mark:

**Comment by leecalcote** [comment]
@Chadha93 tests not passing, it seems...

File: ui/utils/camelCaseToSentanceCase.js [hunk 5]
```diff
@@ -1,3 +1,3 @@
-export const CamelCaseToSentanceCase = (stringValue) => {
+export const CamelCaseToSentenceCase = (stringValue) => {
```
Comment (vaibhav135) [inline ui/utils/camelCaseToSentanceCase.js:1]: Suggestion: anything which is not a component or enum, usually should not be starting with an upper case.

File: .github/workflows/jest-unit-tests.yml [hunk 6]
```diff
@@ -0,0 +1,21 @@
+name: frontendTestingWithJest
+run-name: ${{ github.actor }} is testing their frontend with Jest
+on:   
+  push:
+    branches:
+      - master
+  pull_request:
+
+jobs:
+  run-jest-frontend-tests:
```
Comment (vaibhav135) [inline .github/workflows/jest-unit-tests.yml:10]: I think just `unit-testing` would be a better name here.

**Review by vaibhav135 — CHANGES_REQUESTED** [review]
Everything else looks good to me.

**Comment by leecalcote** [comment]
@aabidsofi19 what do you think?

**Comment by aabidsofi19** [comment]
@leecalcote  i am strongly thinking to have unittests as move forwards specially for our shared library as it can be breaking point for our all projects . 

Of the  test framework i am thinking to go for vittest  it is really fast and elegant ( which means more tests will written and also frequently run ) . rather than just test for pure functions i am also thinking to include component based testing ( vite supports it also ) .  

As we do proper unit and component based testing ( sort of integration tests) we might not need e2e tests for our design system and very little for the overall projects ( following the testing pyramid)

**Comment by leecalcote** [comment]
> @leecalcote i am strongly thinking to have unittests as move forwards specially for our shared library as it can be breaking point for our all projects .
> 
> Of the test framework i am thinking to go for vittest it is really fast and elegant ( which means more tests will written and also frequently run ) . rather than just test for pure functions i am also thinking to include component based testing ( vite supports it also ) .
> 
> As we do proper unit and component based testing ( sort of integration tests) we might not need e2e tests for our design system and very little for the overall projects ( following the testing pyramid)

This sounds promising. I will say, though, that I've been let down by nearly all test frameworks, all of which started off sounding like this. 

How will vittest make tests easier to create, to maintain, and offer unit and integration test coverage? How is vittest better than jest and is this a 5x improvement to cover the switching costs? 

@MarioArriaga92 the clock is ticking on Cypress...

**Comment by aabidsofi19** [comment]
> > @leecalcote i am strongly thinking to have unittests as move forwards specially for our shared library as it can be breaking point for our all projects .
> > Of the test framework i am thinking to go for vittest it is really fast and elegant ( which means more tests will written and also frequently run ) . rather than just test for pure functions i am also thinking to include component based testing ( vite supports it also ) .
> > As we do proper unit and component based testing ( sort of integration tests) we might not need e2e tests for our design system and very little for the overall projects ( following the testing pyramid)
> 
> This sounds promising. I will say, though, that I've been let down by nearly all test frameworks, all of which started off sounding like this.
> 
> How will vittest make tests easier to create, to maintain, and offer unit and integration test coverage? How is vittest better than jest and is this a 5x improvement to cover the switching costs?
> 
> @MarioArriaga92 the clock is ticking on salvaging Cypress...


makes sense , i am proposing it for sistent as we are starting from scratch  so we dont need to rewrite anything
